src/core/dfn-panel.js

```diff
@@ -0,0 +1,161 @@
+// Constructs "dfn panels" which show all the local references to a dfn and a
+// self link to the selected dfn. Based on Bikeshed's dfn panels at
+// https://github.com/tabatkins/bikeshed/blob/ef44162c2e/bikeshed/dfnpanels.py
+import { fetchAsset } from "./text-loader.js";
+import { hyperHTML } from "./import-maps.js";
+import { norm } from "./utils.js";
+
+const cssPromise = loadStyle();
+
+export const name = "core/dfn-panel";
+
+export async function run() {
+  const css = await cssPromise;
+  document.head.insertBefore(
+    hyperHTML`<style>${css}</style>`,
+    document.querySelector("link")
+  );
+
+  document.body.addEventListener("click", event => {
+    /** @type {HTMLElement} */
+    const el = event.target;
+    const panel = document.getElementById("dfn-panel");
+
+    const action = parseAction(el);
+    switch (action) {
+      case "SHOW": {
+        if (panel) panel.remove();
+        const dfn = el.closest("dfn");
+        displayPanel(dfn, createPanel(dfn));
+        break;
+      }
+      case "ACTIVATE": {
+        panel.classList.add("activated");
+        panel.style.left = null;
+        panel.style.top = null;
+        break;
+      }
+      case "HIDE": {
+        panel.remove();
+        break;
+      }
+    }
+  });
+}
+
+/** @param {HTMLElement} clickTarget */
+function parseAction(clickTarget) {
+  const hitALink = !!clickTarget.closest("a");
+  if (clickTarget.closest("dfn")) {
+    return hitALink ? "NOOP" : "SHOW";
+  }
+  if (clickTarget.closest("#dfn-panel")) {
+    if (hitALink) {
+      const clickedSelfLink = clickTarget.classList.contains("self-link");
+      return clickedSelfLink ? "HIDE" : "ACTIVATE";
+    }
+    const panel = clickTarget.closest("#dfn-panel");
+    return panel.classList.contains("activated") ? "HIDE" : "NOOP";
+  }
+  if (document.getElementById("dfn-panel")) {
+    return "HIDE";
+  }
+  return "NOOP";
+}
+
+/** @param {HTMLElement} dfn */
+function createPanel(dfn) {
+  const { id } = dfn;
+  const href = `#${id}`;
+  const links = document.querySelectorAll(`a[href="${href}"]`);
+
+  /** @type {HTMLElement} */
+  const panel = hyperHTML`
+    <aside class="dfn-panel" id="dfn-panel">
+      <b><a class="self-link" href="${href}">${href}</a></b>
+      ${
+        links.length
+          ? hyperHTML`<b>Referenced in:</b>${referencesToHTML(id, links)}`
+          : null
+      }
+    </aside>
+  `;
+  return document.body.appendChild(panel);
+}
+
+/**
+ * @param {string} id dfn id
+ * @param {NodeListOf<HTMLLinkElement>} links
+ * @returns {HTMLUListElement}
+ */
+function referencesToHTML(id, links) {
+  /** @type {Map<string, string[]>} */
+  const titleToIDs = new Map();
+  links.forEach((link, i) => {
+    const linkID = link.id || `ref-for-${id}-${i + 1}`;
+    if (!link.id) link.id = linkID;
```

Comment:
Maybe do this before... then you don't need to build linkID if it's not going to be used?